src/main/java/org/tron/common/overlay/portmap/model/PortMappingPreset.java

```diff
@@ -0,0 +1,173 @@
+ *
+ * You should have received a copy of the GNU General Public License along with this program.  If
+ * not, see <http://www.gnu.org/licenses/>.
+ */
+package org.tron.common.overlay.portmap.model;
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+import org.tron.common.overlay.portmap.Settings;
+
+/**
+ * This class stores a port mapping preset containing a description, internal and remote host and a
+ * {@link List} of {@link SinglePortMapping}s.
+ */
+public class PortMappingPreset implements Cloneable, Serializable {
+
+  private static final long serialVersionUID = 3749136884938395765L;
+
+  /**
+   * The description of this preset.
+   */
+  private String description;
+
+  /**
+   * The ip address of the internal client or <code>null</code> for localhost.
+   */
+  private String internalClient;
+
+  /**
+   * The host name of the remote host.
+   */
+  private String remoteHost;
+
+  /**
+   * The port mappings in this preset.
+   */
+  private List<SinglePortMapping> ports;
+
+  /**
+   * <code>true</code> if this preset has not been saved.
+   */
+  private boolean isNew;
+
+  /**
+   * Creates a new preset with the given default values.
+   */
+  public PortMappingPreset(final String remoteHost, final String internalClient,
+      final String description) {
+    this.remoteHost = remoteHost;
+    this.internalClient = internalClient;
+    this.description = description;
+    this.ports = new LinkedList<>();
+
+    this.isNew = false;
+  }
+
+  /**
+   * Creates a new empty preset.
+   */
+  public PortMappingPreset() {
+    this.ports = new LinkedList<>();
+    this.isNew = true;
+  }
+
+  @Override
+  public String toString() {
+    return description;
+  }
+
+  public List<PortMapping> getPortMappings(final String localhost) {
+    if (this.useLocalhostAsInternalClient() && (localhost == null || localhost.length() == 0)) {
+      throw new IllegalArgumentException("Got invalid localhost and internal host is not given.");
+    }
+
+    final List<PortMapping> allPortMappings = new ArrayList<>(this.ports.size());
+    for (final SinglePortMapping port : this.ports) {
+      final String internalClientName =
+          this.useLocalhostAsInternalClient() ? localhost : this.internalClient;
+
+      final PortMapping newMapping = new PortMapping(port.getProtocol(), remoteHost,
+          port.getExternalPort(),
+          internalClientName, port.getInternalPort(), description);
+
+      allPortMappings.add(newMapping);
+    }
+
+    return allPortMappings;
+  }
+
+  public String getCompleteDescription() {
+    final StringBuffer b = new StringBuffer();
+
+    b.append(" ");
+    b.append(remoteHost);
+    b.append(":");
+
+    b.append(" -> ");
+    b.append(internalClient);
+    b.append(":");
+
+    b.append(" ");
+
+    b.append(" ");
+    b.append(description);
+    return b.toString();
+  }
+
+  public List<SinglePortMapping> getPorts() {
+    return ports;
```

Comment:
![Vulnerability](https://sonarsource.github.io/sonar-github/vulnerability.png 'Vulnerability') Vulnerability: Return a copy of "ports". ([squid:S2384](https://sonarcloud.io/organizations/tron-zhaohong/rules#rule_key=squid%3AS2384))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=java-tron&pullRequest=1583&issues=AWZcuxqi9tx2us2yFA5Z&open=AWZcuxqi9tx2us2yFA5Z)

[ISSUE_KEY]: AWZcuxqi9tx2us2yFA5Z